Estimate the resources needed for a VariableResources task.

nh13 opened the pull request:
This is used when a VariableResources task cannot be scheduled, and we'd like to know the minimum amount of resources it needs.

@tfenne I should add some tests, but I wanted to see what you thought.  I could also try a search on the diagonal too, but I thought that was overkill.

tfenne commented on core/src/main/scala/dagr/core/tasksystem/Schedulable.scala:
```diff
@@ -89,4 +89,30 @@ trait FixedResources extends ScheduleWithEmptyDefaultResources {
+  def minCoresAndMemory(maxCores: Cores = Cores(32), maxMemory: Memory = Memory("64g")): ResourceSet = {
+    val cores  = minCores(end=maxCores).getOrElse(Cores.infinite)
+    val memory = minMemory(end=maxMemory).getOrElse(Memory.infinite)
+    new ResourceSet(cores, memory)
```
Perhaps at least test that if you offer it `cores` and `memory` after computing them separately, that that's sufficient?

nh13 commented:
@tfenne made some changes you suggested and more.  I added in the ability to search using per-core memory ratio, which is common across my pipelines.  I moved things up into `Schedulable` too.  Could you take another look?

tfenne commented on core/src/main/scala/dagr/core/tasksystem/Schedulable.scala:
```diff
@@ -48,6 +48,65 @@ trait Schedulable {
+    (cores, memory) match {
+      case (Some(c), Some(m)) => this.pickResources(maximumResources.copy(cores=c, memory=m))
+      case _                  =>
+        // try a combination of cores and memory with a fixed reatio between the two.
```
reatio -> ratio

tfenne commented on core/src/main/scala/dagr/core/tasksystem/Schedulable.scala:
```diff
@@ -48,6 +48,65 @@ trait Schedulable {
+    * @param maximumResources The maximum system resources available to the task.
+    * @return Either a ResourceSet of the subset of resources that this task can run with, or None
+    */
+  private[core] def minResources(maximumResources: ResourceSet = new ResourceSet(cores=Cores(32), memory=Memory("64g"))): Option[ResourceSet] = {
```
The defaults here feel very "no-one will ever need more than 640kb of memory"-esque.  I would either set these way higher (like 1PB and 1000 cores) or set the as some very large multiple of the current system resources.

tfenne commented on core/src/main/scala/dagr/core/tasksystem/Schedulable.scala:
```diff
@@ -48,6 +48,65 @@ trait Schedulable {
+        // try a combination of cores and memory with a fixed reatio between the two.
+        val startMemoryPerCore = Memory("256m").value
+        val endMemoryPerCore   = maximumResources.memory.value
+        val stepMemoryPerCore  = Memory("256m").value
```
Instead of hard-coding this to 256m, could you set it to be some reasonable fraction (e.g. 1/100th) of the max memory so you have a fixed number of steps?

nh13 commented:
@tfenne I made the search based on a multiple of the system resources.  If we find that these estimates don't yield good results, we can always improve them.  Lets table that discussion since I have a lot of complexity in their for now.

tfenne commented:
Failing tests need fixing, then 👍
